### ISSUE TITLE
Timed-out requests never return

### ISSUE THREAD

**Issue by jonathanstrong** [sagebind/chttp#3]
I set up a `Client` like this:

```rust
extern crate chttp;
use std::time::Duration;

fn http_client() -> Client {
    let mut opt = chttp::Options::new();
    opt.timeout = Some(Duration::from_secs(2));
    opt.connect_timeout = Duration::from_secs(2);
    Client::builder()
        .max_connections(Some(4))
        .options(opt)
        .build()
}
```

I am sending a post request to localhost:8080, where I have a simple python http server listening on 8080 that sleeps 3 seconds before responding (> 2s timeout).

I create the `Request<Body>` and send via `Client::send`. The request is immediately registered by the http server, which echos the raw http request to the console:

```console
Host: localhost:8080
# [custom header omitted]
# [custom header omitted]
content-type: application/x-www-form-urlencoded
content-length: 53

# [53-length post data omitted]
```

About three seconds later, the http server registers a broken connection...

```console
127.0.0.1 - - [02/Jul/2018 17:51:20] "POST / HTTP/1.1" 200 -
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 54498)
Traceback (most recent call last):
  File "/usr/lib/python2.7/SocketServer.py", line 290, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 318, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 331, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python2.7/SocketServer.py", line 654, in __init__
    self.finish()
  File "/usr/lib/python2.7/SocketServer.py", line 713, in finish
    self.wfile.close()
  File "/usr/lib/python2.7/socket.py", line 283, in close
    self.flush()
  File "/usr/lib/python2.7/socket.py", line 307, in flush
    self._sock.sendall(view[write_offset:write_offset+buffer_size])
error: [Errno 32] Broken pipe
----------------------------------------
```

...but the rust program waiting on the `Client::send` call hangs indefinitely.

Am I doing anything wrong? I expected the `Client::send` call to return an error immediately when the timeout was reached.

**Comment by jonathanstrong** [comment]
here is my curl version, if that helps:

```console
$ curl --version
curl 7.47.0 (x86_64-pc-linux-gnu) libcurl/7.47.0 GnuTLS/3.4.10 zlib/1.2.8 libidn/1.32 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets 
```

**Comment by sagebind** [comment]
Your program seems correct, this must be a bug somewhere, whether in libcurl or in cHTTP. I will look into this further.

Thanks for letting me know of the issue!

**Comment by jonathanstrong** [comment]
any luck hunting this down?

I have been using hyper v0.10 as I did not want to go down the async rabbit hole, and originally turned to this library for decent timeout behavior (heh). But I also benchmarked the two libraries against each other today and chttp is substantially faster in my program. Would love to use it but the timeouts issue is a problem.

I would offer to help but I doubt it would be of much assistance, as I don't know C and have zero familiarity with the libcurl codebase.

Thanks for any info/help you can provide, and for making the library available in the first place!

**Comment by sagebind** [comment]
Glad it is useful!

I've been in a bit of a time-crunch lately and this fell on the backburner. I'd definitely like to look into this, but I can't guarantee a timeline.

I also hope to do some internal refactoring to use the cURL handles more efficiently to make it match bare metal cURL speeds (there's a small overhead I think).

If you are able, it would help to see if you could reproduce this problem by using the `curl` crate directly using the "curl easy" API in there. That would indicate whether the problem is in this crate or in cURL (or the `curl` crate).

**Comment by jonathanstrong** [comment]
sure thing - will investigate and report back, thanks.

**Comment by jonathanstrong** [comment]
here is a self-contained crate to test this: https://github.com/jonathanstrong/curl-timeout

using the `Easy` api, the request does time out successfully on my machine (debug build). I'm sure there's more hacking in terms of the different types of timeouts available but this is a good platform to evaluate it with, I think.

**Comment by sagebind** [comment]
@jonathanstrong That's a good start, thanks. The bug seems to be somewhere in this crate then. Possibly related to synchronization issues. I'll see if I can look into this this week.

**Comment by sagebind** [comment]
@jonathanstrong Great news, I found the bug! I have implemented a bugfix in #4, which I will merge in and publish shortly.

In layman's terms, libcurl did in fact respect the timeout, but discarded the notification that this happened before chttp could read it and update its state, so the client kept looping indefinitely, waiting for a response. This happened because of a misuse of the libcurl API (calling functions in the wrong order).
